### PR TITLE
[Major] Removed Material UI dependencies from Dropdown

### PR DESCRIPTION
**Pull request by GomezIvann**
Summary of behaviour changes of the component regarding the original:

1. New `box-shadow` to match other components with a list of Halstack.
2. Dropdown menu has been reimplemented as a Portal using Radix UI Popover, similar to how it was done with the select.
3. Accessibility roles and attributes are preserved as much as possible, adding new ones following [W3C guidelines](https://www.w3.org/WAI/ARIA/apg/example-index/menu-button/menu-button-actions-active-descendant.html) for the active descendant pattern.
5. Menu transitions for opening and closing have been removed.
6. Keyboard Support has been preserved, adding new ones such as `Esc`, `Home`, `PageUp`, `End`, `PageDown`, etc. For this, I've followed W3C accessibility guidelines for Menu Buttons (Dropdown pattern).
7. Storybooks and Jest tests have been rewritten and incremented to match the new implementation and catch possible future errors.

### CODE REVIEW

File: lib/src/dropdown/Dropdown.test.js [hunk 1]
```diff
@@ -1,145 +1,282 @@
...
-    const { queryByText } = render(
-      <DxcDropdown options={options} expandOnHover={true} label="dropdown-test"></DxcDropdown>
+
+    fireEvent.keyDown(getByRole("button"), { key: "ArrowUp", code: "ArrowUp", keyCode: 38, charCode: 38 });
+    const menu = getByRole("menu");
+    fireEvent.keyDown(menu, { key: "ArrowDown", code: "ArrowDown", keyCode: 40, charCode: 40 });
+    expect(document.activeElement === menu).toBeTruthy();
+    expect(menu.getAttribute("aria-activedescendant")).toBe("option-0");
+    fireEvent.keyDown(menu, { key: "Enter", code: "Enter", keyCode: 13, charCode: 13 });
+    expect(onSelectOption).toHaveBeenCalledWith("1");
```
Comment (marcialfps) [inline lib/src/dropdown/Dropdown.test.js:215]: Why is it called just one time? 🤔 I think I'm losing something...

File: lib/src/dropdown/Dropdown.stories.tsx [hunk 2]
```diff
@@ -3,21 +3,34 @@ import { userEvent, within } from "@storybook/testing-library";
...
+import DropdownMenu from "./DropdownMenu";
+import DxcCheckbox from "../checkbox/Checkbox";
+import DxcTextInput from "../text-input/TextInput";
 
 export default {
   title: "Dropdown",
   component: DxcDropdown,
 };
 
+type Option = {
```
Comment (marcialfps) [inline lib/src/dropdown/Dropdown.stories.tsx:15]: Maybe you can import the type from `types.ts`.

**Review by marcialfps — CHANGES_REQUESTED** [review]

File: lib/src/dropdown/Dropdown.tsx [hunk 3]
```diff
@@ -20,145 +29,190 @@ const DxcDropdown = ({
...
+        case "ArrowUp":
+          event.preventDefault();
+          setPreviousIndexFocus();
+          break;
+        case "Down":
+        case "ArrowDown":
+          event.preventDefault();
+          setNextIndexFocus();
+          break;
+        case "Space":
```
[the remark this answers is not in the document]
Comment (GomezIvann) [reply lib/src/dropdown/Dropdown.tsx:114]: [W3C guidelines](https://www.w3.org/WAI/ARIA/apg/example-index/menu-button/menu-button-links.html) give the Space key the same behaviour as the Enter key 🤔

File: lib/src/dropdown/Dropdown.tsx [hunk 4]
```diff
@@ -20,145 +29,190 @@ const DxcDropdown = ({
...
-    if (ref.current) setWidth(ref.current.offsetWidth);
+  const handleOnOpenMenu = () => {
+    changeIsOpen(true);
+  };
+  const handleOnCloseMenu = () => {
+    changeIsOpen(false);
+    setVisualFocusIndex(0);
+  };
+  const handleMenuItemOnClick = useCallback(
+    (option) => {
```
[the remark this answers is not in the document]
Comment (GomezIvann) [reply lib/src/dropdown/Dropdown.tsx:57]: I don't think it would make any difference but I changed it if you think It makes the code more readable 👍🏼.

File: lib/src/dropdown/Dropdown.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (GomezIvann) [reply lib/src/dropdown/Dropdown.tsx:114]: Look at their actual [implementation](https://codepen.io/pen).

File: lib/src/dropdown/Dropdown.tsx [hunk 5]
```diff
@@ -20,145 +29,190 @@ const DxcDropdown = ({
...
+          handleMenuItemOnClick(options[visualFocusIndex]);
+          break;
+        case "Esc":
+        case "Escape":
+          event.preventDefault();
+          handleOnCloseMenu();
+          triggerRef.current?.focus();
+          break;
+        case "Home":
+        case "PageUp":
```
[the remark this answers is not in the document]
Comment (GomezIvann) [reply lib/src/dropdown/Dropdown.tsx:126]: Same as above, not documented but their are in their implementation (which I looked while implementing it).

File: lib/src/dropdown/Dropdown.test.js [hunk 6]
```diff
@@ -1,145 +1,282 @@
...
+    expect(dropdown.getAttribute("aria-disabled")).toBe("true");
+    expect(queryByRole("menu")).toBeFalsy();
+    userEvent.click(dropdown);
+    expect(queryByRole("menu")).toBeFalsy();
+    expect(queryByText("Amazon")).toBeFalsy();
+    userEvent.click(dropdown);
+    expect(queryByRole("menu")).toBeFalsy();
+    expect(dropdown.getAttribute("aria-expanded")).toBe("false");
+  });
+  test("onSelectOption function is called correctly when an option is called", () => {
```
[the remark this answers is not in the document]
Comment (GomezIvann) [reply lib/src/dropdown/Dropdown.test.js:85]: Done!

File: lib/src/dropdown/Dropdown.test.js [hunk 7]
```diff
@@ -1,145 +1,282 @@
...
+    );
+    const dropdown = getByRole("button");
+
     expect(queryByText("option-test")).toBeFalsy();
+    expect(queryByRole("menu")).toBeFalsy();
+    fireEvent.mouseOver(dropdown);
+    const menu = getByRole("menu");
+    expect(menu).toBeTruthy();
+    expect(document.activeElement === menu).toBeTruthy();
+    expect(getByRole("menu").getAttribute("aria-activedescendant")).toBe("option-0");
```
[the remark this answers is not in the document]
Comment (GomezIvann) [reply lib/src/dropdown/Dropdown.test.js:109]: Corrected!

File: lib/src/dropdown/Dropdown.test.js [hunk 8]
```diff
@@ -1,145 +1,282 @@
...
+    userEvent.click(option);
+    expect(onSelectOption).toHaveBeenCalledWith("4");
+  });
+  test("When expandOnHover is true, the dropdown trigger shows and hides the menu when it is hovered", () => {
+    const { queryByText, getByRole, queryByRole } = render(
+      <DxcDropdown options={options} expandOnHover label="dropdown-test" />
+    );
+    const dropdown = getByRole("button");
+
     expect(queryByText("option-test")).toBeFalsy();
```
[the remark this answers is not in the document]
Comment (GomezIvann) [reply lib/src/dropdown/Dropdown.test.js:103]: Corrected!

File: lib/src/dropdown/Dropdown.test.js [hunk 9]
```diff
@@ -1,145 +1,282 @@
...
-  test("Dropdown renders with caret hidden", () => {
-    const { queryByRole } = render(<DxcDropdown caretHidden={true} label="dropdown-test"></DxcDropdown>);
-    expect(queryByRole("img")).toBeFalsy();
+    userEvent.click(getByRole("button"));
+    const menu = getByRole("menu");
+    fireEvent.keyDown(menu, { key: "ArrowUp", code: "ArrowUp", keyCode: 38, charCode: 38 });
+    expect(document.activeElement === menu).toBeTruthy();
+    expect(menu.getAttribute("aria-activedescendant")).toBe("option-3");
+    fireEvent.keyDown(menu, { key: "Enter", code: "Enter", keyCode: 13, charCode: 13 });
+    expect(onSelectOption).toHaveBeenCalledWith("4");
```
[the remark this answers is not in the document]
Comment (GomezIvann) [reply lib/src/dropdown/Dropdown.test.js:186]: `toHaveBeenCalledWith` is a jest function that checks the parameter of a tested function. In this case, I'm checking that the Enter key calls the `onSelectOption` function with the focused option "Aliexpress" which is "4".

File: lib/src/dropdown/Dropdown.tsx [hunk 10]
```diff
@@ -20,145 +29,190 @@ const DxcDropdown = ({
...
-          </DXCMenu>
-        </div>
-      </DXCDropdownContainer>
+      <DropdownContainer
+        onMouseEnter={!disabled && expandOnHover ? handleOnOpenMenu : undefined}
+        onMouseLeave={!disabled && expandOnHover ? handleOnCloseMenu : undefined}
+        onBlur={!disabled ? handleOnBlur : undefined}
+        margin={margin}
+        size={size}
+      >
```
[the remark this answers is not in the document]
Comment (GomezIvann) [reply lib/src/dropdown/Dropdown.tsx:164]: I tried before and It didn't work as intended, let me try again.

[the remark this answers is not in the document]
Comment (GomezIvann) [reply lib/src/dropdown/Dropdown.tsx:164]: Yes, I confirm. This cannot be achieved for some reasons:

1. The focus here is a little bit trickier than in other components such as Select or Autosuggest. There, the focus always stays on the trigger that opens the list, but in the dropdown that is not correct. The focus moves between the list and the button when opening and closing, and should go out when clicked outside of them. Since we are using portals for the menu this becomes more difficult because the menu is not inside the main container, so I needed to play with the events, their propagation, etc. I need the outside container so the trigger stops its onBlur propagation (when the button loses the focus to the menu it should not close it).
2. `expandOnHover` is another issue here. If we land the `mouseEnter` and `mouseLeave` events in the trigger instead of the container the bubbling would close the menu when the cursor goes from the trigger to the menu.

File: lib/src/dropdown/Dropdown.tsx [hunk 11]
```diff
@@ -20,145 +29,190 @@ const DxcDropdown = ({
...
+        case "End":
+        case "PageDown":
+          event.preventDefault();
+          setVisualFocusIndex(options.length - 1);
+          break;
+        case "Tab":
+          handleOnCloseMenu();
+          triggerRef.current?.focus();
+          break;
+      }
```
[the remark this answers is not in the document]
Comment (GomezIvann) [reply lib/src/dropdown/Dropdown.tsx:139]: I think I'm going to postpone this functionality, since it would require adding it in other not-as-easy components, such as select.

File: lib/src/dropdown/Dropdown.tsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (marcialfps) [reply lib/src/dropdown/Dropdown.tsx:170]: Okay! Nice explanation Ivan! 💯

File: lib/src/dropdown/Dropdown.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (marcialfps) [reply lib/src/dropdown/Dropdown.tsx:115]: In [Menu pattern ](https://www.w3.org/WAI/ARIA/apg/patterns/menu/) they mark this functionality as optional:

> (Optional): When focus is on a menuitem that does not have a submenu, activates the menuitem and closes the menu.

I think we can maintain it. 👍🏼

File: lib/src/dropdown/Dropdown.tsx (hunk 11 above)
[the remark this answers is not in the document]
Comment (marcialfps) [reply lib/src/dropdown/Dropdown.tsx:140]: Maybe should we open an issue?

File: lib/src/dropdown/Dropdown.test.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (marcialfps) [reply lib/src/dropdown/Dropdown.test.js:186]: Sorry! I read `toHaveBeenCalledTimes`. 🤦🏼‍♂️

File: lib/src/dropdown/Dropdown.test.js (hunk 1 above)
Comment (marcialfps) [reply lib/src/dropdown/Dropdown.test.js:215]: 🤦🏼‍♂️

**Review by marcialfps — APPROVED** [review]
💯

**Review by marcialfps — APPROVED** [review]

File: lib/src/dropdown/Dropdown.tsx (hunk 11 above)
[the remark this answers is not in the document]
Comment (GomezIvann) [reply lib/src/dropdown/Dropdown.tsx:140]: Maybe an issue is a bit misleading, this should be considered as an enhancement worth of discussing. What do you think?
